### PR TITLE
Initial set of pillar types related files for cluster multi-node mode

### PR DESCRIPTION
**Pull request by naiming-zededa**
- for cluster multi-node mode, some of the initial pillar side types related files, additional changes will be added in later PRs

### CODE REVIEW

File: pkg/pillar/types/clustertypes.go [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+
+const (
+	ClusterStatusPort = "12346" // ClusterStatusPort - Port for k3s server for cluster status advertise
+)
+
+// EdgeNodeClusterConfig - Configuration for cluster multi-node from controller
+type EdgeNodeClusterConfig struct {
+	ClusterName      string
+	ClusterID        UUIDandVersion
+	ClusterInterface string
```
Comment (milan-zededa) [inline pkg/pillar/types/clustertypes.go:20]: Add comment that this is a logical label of the port.

File: pkg/pillar/types/clustertypes.go [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+	ClusterIPPrefix  *net.IPNet
+	IsWorkerNode     bool
+	JoinServerIP     net.IP
+	BootstrapNode    bool
```
Comment (milan-zededa) [inline pkg/pillar/types/clustertypes.go:24]: We should also add comment explaining what "bootstrap" means in this context.

File: pkg/pillar/types/clustertypes.go [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+// Key - returns the key for the config of EdgeNodeClusterConfig
+func (config EdgeNodeClusterConfig) Key() string {
+	return config.ClusterID.UUID.String()
+}
+
+// EdgeNodeClusterStatus - Status of the multi-node cluster published by zedkube
+type EdgeNodeClusterStatus struct {
+	ClusterName      string
+	ClusterID        UUIDandVersion
+	ClusterInterface string
```
Comment (milan-zededa) [inline pkg/pillar/types/clustertypes.go:48]: Add comment that this is a logical label of the port.

File: pkg/pillar/types/clustertypes.go [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+	BootstrapNode    bool
+
+	// CipherBlockStatus, for encrypted cluster token data
+	CipherToken CipherBlockStatus
+}
+
+// ENClusterAppStatus - Status of an App Instance in the multi-node cluster
+type ENClusterAppStatus struct {
+	AppUUID             uuid.UUID // UUID of the appinstance
+	IsDNSet             bool      // DesignatedNodeID is set for this node
```
Comment (milan-zededa) [inline pkg/pillar/types/clustertypes.go:33]: In `DomainConfig` you use `DNidNode`, should we use the same name for this variable as well? Or does `DN` and `DNid` refer to different things?

File: pkg/pillar/types/domainmgrtypes.go [hunk 5]
```diff
@@ -474,6 +481,7 @@ type DomainMetric struct {
 	UsedMemoryPercent float64
 	LastHeard         time.Time
 	Activated         bool
+	NodeName          string // kube app running on node
```
Comment (milan-zededa) [inline pkg/pillar/types/domainmgrtypes.go:484]: Maybe more clear comment would be: "the name of the kubernetes node on which the app is running"

File: pkg/pillar/types/zedkubetypes.go [hunk 6]
```diff
@@ -0,0 +1,83 @@
...
+	KubeNodeStatusNotReady                           // KubeNodeStatusNotReady - Node is in not ready status
+	KubeNodeStatusNotReachable                       // KubeNodeStatusNotReachable - Node is not reachable
+)
+
+// KubeNodeInfo - Information about a Kubernetes node
+type KubeNodeInfo struct {
+	Name               string
+	Status             KubeNodeStatus
+	IsMaster           bool
+	IsEtcd             bool
```
Comment (milan-zededa) [inline pkg/pillar/types/zedkubetypes.go:23]: How about `UsesEtcd`?

File: pkg/pillar/types/zedkubetypes.go [hunk 7]
```diff
@@ -0,0 +1,83 @@
...
+	CreationTime       time.Time
+	LastTransitionTime time.Time
+	KubeletVersion     string
+	InternalIP         string
```
Comment (milan-zededa) [inline pkg/pillar/types/zedkubetypes.go:27]: I would suggest using net.IP type for IP addresses.

File: pkg/pillar/types/zedkubetypes.go [hunk 8]
```diff
@@ -0,0 +1,83 @@
...
+)
+
+// KubePodInfo - Information about a Kubernetes pod
+type KubePodInfo struct {
+	Name              string
+	Status            KubePodStatus
+	RestartCount      int32
+	RestartTimestamp  time.Time
+	CreationTimestamp time.Time
+	PodIP             string
```
[the remark this answers is not in the document]
Comment (naiming-zededa) [reply pkg/pillar/types/zedkubetypes.go:50]: For that, we are only 'relaying' from the kubernetes `type PodStatus struct`, and the `PodIP` is a string there.
```
	PodIP string `json:"podIP,omitempty" protobuf:"bytes,6,opt,name=podIP"`
```

File: pkg/pillar/types/zedkubetypes.go (hunk 7 above)
Comment (naiming-zededa) [reply pkg/pillar/types/zedkubetypes.go:27]: We got this from kubernetes 'nodestatus' struct, which is a list of string for 'Adddress'.

File: pkg/pillar/types/zedkubetypes.go (hunk 6 above)
Comment (naiming-zededa) [reply pkg/pillar/types/zedkubetypes.go:23]: Sure.

File: pkg/pillar/types/domainmgrtypes.go (hunk 5 above)
Comment (naiming-zededa) [reply pkg/pillar/types/domainmgrtypes.go:484]: Sure.

File: pkg/pillar/types/clustertypes.go (hunk 4 above)
Comment (naiming-zededa) [reply pkg/pillar/types/clustertypes.go:33]: Same meaning. I'll change this to IsDNidNode here.

File: pkg/pillar/types/clustertypes.go (hunk 3 above)
Comment (naiming-zededa) [reply pkg/pillar/types/clustertypes.go:48]: Ok.

File: pkg/pillar/types/clustertypes.go (hunk 2 above)
Comment (naiming-zededa) [reply pkg/pillar/types/clustertypes.go:24]: Sure.

File: pkg/pillar/types/clustertypes.go (hunk 1 above)
Comment (naiming-zededa) [reply pkg/pillar/types/clustertypes.go:20]: OK.

File: pkg/pillar/types/clustertypes.go [hunk 9]
```diff
@@ -0,0 +1,56 @@
...
+package types
+
+import (
+	"net"
+
+	uuid "github.com/satori/go.uuid"
+)
+
+const (
+	ClusterStatusPort = "12346" // ClusterStatusPort - Port for k3s server for cluster status advertise
```
[the remark this answers is not in the document]
Comment (naiming-zededa) [reply pkg/pillar/types/clustertypes.go:13]: Will have more docs in pillar/docs/zedkube.md for this item. and i can mention this doc here.

File: pkg/pillar/types/zedkubetypes.go (hunk 7 above)
Comment (milan-zededa) [reply pkg/pillar/types/zedkubetypes.go:27]: OK

File: pkg/pillar/types/zedkubetypes.go (hunk 8 above)
[the remark this answers is not in the document]
Comment (milan-zededa) [reply pkg/pillar/types/zedkubetypes.go:50]: OK

**Review by milan-zededa — APPROVED** [review]
All my comments have been addressed

**Comment by naiming-zededa** [comment]
@eriknordmark can we merge this in and we will have other cluster related PRs depends on this one.

**Pull request merged**